_posts/2016-03-16-AngularTS.md

```diff
@@ -0,0 +1,413 @@
+       
+              }
+              getEmployeeByLink(href:string):ng.IPromise<{}>{
+                  return this.employees.oneUrl(href).get();
+              }       
+          }
+          angular.module("jworks360")
+              .service("EmployeeService", EmployeeService);       
+       }
+{% endhighlight %}
+
+In the above example, restangular has its own tsd file to reference to.
+For services it is best practice to declare an interface for data typing and get a clear view of the intent.
+The service class will implement all methods related to the data communication with the backend and returns a promise to the controllers or services that will inject this custom service.
+RestAngular has it’s own configuration you can modify in the .config component to point to the right api call.
+After the config you can inject the restangular service and use its services to build up a request to the backend.
+
+
+## Directives
+Custom directives allow you to create highly semantic and reusable components.
+A directive allows angular to manipulate the DOM and add it’s own behaviours. 
+These can be either a set of instructions or a JSON representation.
+Defining a directive in TypeScript we use the directive service angular provides. 
+
+{% highlight javascript %}
+module JWorks {
+
+   export interface IAnimate extends ng.IAttributes {
+       jwAnimate:string;
+   }
+
+   class Animate implements ng.IDirective {
+
+       restrict = "A";
+
+       static instance():ng.IDirective {
+           return new Animate();
+       }
+
+       link($scope, elm:ng.IRootElementService, attr:IAnimate,ngModel:ng.INgModelController):void {
+
+           $scope.right = function(){
+               $(this).animate({
+                   left: '+=150'
+               });
+                elm.fadeOut("slow");
+           };
+           var direction = attr['jwAnimate'];
+           elm.on('click',$scope[direction]);
+
+       }
+
+   }
+   angular.module("jworks360").directive("jwAnimate", Animate.instance);
+
+}
+{% endhighlight %}
+
+In the interface above we have to tell angular what name we will use for our directive.
+The attribute service will be called to add an additional string to the attribute service.
+Secondly the class has to implement the directive interface to be recognized by the compiler as a directive.
+Inside the class you have to declare the prefixed properties and override the methods you will be using.
+The ‘static instance()’ method has to be declared to let your module know that there is a new directive.
+At the end you register the directive to your module with the instance as value.
+
+
+
+## Final note
+
+Best practices can change over time. 
+With webpack for example the registry to the module is gathered in one file.
+TypeScript keeps on growing and in my opinion will be the default language for many future front-end projects.
+When we arrive at testing, TypeScript provides better support because of encapsulation. 
```

Comment:
Reword to:
When **it comes to testing our code**, TypeScript **will provide** better ...